File: tests_end_to_end/page_objects/ThreadsPage.py
import time
from playwright.sync_api import Page, expect


class ThreadsPage:
    def __init__(self, page: Page):
        self.page = page
        self.search_input = self.page.get_by_test_id("search-input")
        self.threads_tab = self.page.get_by_role("tab", name="Threads")
        self.thread_row = self.page.locator("tbody tr")
        self.thread_container = self.page.get_by_test_id("thread")
        self.thread_container_delete_button = self.page.get_by_role(
            "button", name="Delete"
        )
        self.thread_delete_popup_delete_button = self.page.get_by_role("button").filter(
            has_text="Delete thread"
        )
        self.thread_checkbox = self.page.get_by_label("Select row")
        self.thread_table_delete_button = (
            self.page.get_by_role("tabpanel")
            .locator("div")
            .filter(has_text="Columns")
            .get_by_role("button")
            .nth(3)
        )
        self.output_container = self.page.get_by_test_id("thread").locator(
            ".comet-markdown"
        )

    def switch_to_page(self):
        self.threads_tab.click()

    def get_number_of_threads_on_page(self):
        try:
            for i in range(5):
                if not self.thread_row.count():
                    time.sleep(1)
                    self.page.reload()
            expect(self.thread_row.first).to_be_visible()
        except Exception as e:
            raise AssertionError(
                f"No threads found in the project.\n"
                f"Error: {str(e)}"
            ) from e
        finally:
            return self.thread_row.count()

    def open_thread_content(self, thread_id):
        self.page.get_by_role("button", name=thread_id).click()

    def check_message_in_thread(self, message, output=False):
        if output:
            expect(self.output_container.filter(has_text=message)).to_be_visible()
        else:
            expect(self.thread_container.get_by_text(message)).to_be_visible()

    def search_for_thread(self, thread_id):
        self.search_input.fill(thread_id)
        expect(self.thread_row).to_have_count(1)

    def delete_thread_from_table(self):
        self.thread_checkbox.click()
        self.thread_table_delete_button.click()
        self.thread_delete_popup_delete_button.click()

    def check_thread_is_deleted(self, thread_id):
        self.search_input.fill(thread_id)
        expect(self.thread_row.filter(has_text=thread_id)).to_have_count(0)

    def delete_thread_from_thread_content_bar(self):
        self.thread_container_delete_button.click()
        self.thread_delete_popup_delete_button.click()

    def close_thread_content(self):
        self.page.keyboard.press("Escape")

File: tests_end_to_end/typescript-tests/page-objects/threads.page.ts
import { Page, Locator, expect } from '@playwright/test';

export class ThreadsPage {
  readonly page: Page;
  readonly searchInput: Locator;
  readonly logsTab: Locator;
  readonly threadsToggle: Locator;
  readonly threadRow: Locator;
  readonly threadContainer: Locator;
  readonly threadContainerDeleteButton: Locator;
  readonly threadDeletePopupDeleteButton: Locator;
  readonly threadCheckbox: Locator;
  readonly threadTableDeleteButton: Locator;
  readonly outputContainer: Locator;

  constructor(page: Page) {
    this.page = page;
    this.searchInput = page.getByTestId('search-input');
    this.logsTab = page.getByRole('tab', { name: 'Logs' });
    this.threadsToggle = page.getByRole('radio', { name: 'Threads' });
    this.threadRow = page.locator('tbody tr');
    this.threadContainer = page.getByTestId('thread');
    this.threadContainerDeleteButton = page
      .locator('div')
      .filter({ hasText: /^Add to/ })
      .getByRole('button')
      .nth(2);
    this.threadDeletePopupDeleteButton = page.getByRole('button', { name: 'Delete' });
    this.threadCheckbox = page.getByLabel('Select row');
    // Find the delete button by its trash icon SVG (same pattern as traces page)
    this.threadTableDeleteButton = page
      .getByRole('button')
      .filter({ has: page.locator('svg path[d*="M3 6h18"]') });
    this.outputContainer = page.getByTestId('thread').locator('.comet-markdown');
  }

  async switchToPage(): Promise<void> {
    // Click the Logs tab first
    await this.logsTab.click();
    // Then click the Threads toggle within the Logs tab
    await this.threadsToggle.click();
  }

  async getNumberOfThreadsOnPage(): Promise<number> {
    try {
      // Retry up to 5 times if no threads found initially
      for (let i = 0; i < 5; i++) {
        const count = await this.threadRow.count();
        if (count === 0) {
          await this.page.waitForTimeout(1000);
          await this.page.reload();
        } else {
          break;
        }
      }

      await expect(this.threadRow.first()).toBeVisible();
      return await this.threadRow.count();
    } catch (error) {
      throw new Error(`No threads found in the project.\nError: ${error}`);
    }
  }

  async openThreadContent(threadId: string): Promise<void> {
    // Click on the row containing the thread ID (not the ID cell itself, which copies to clipboard)
    await this.page
      .getByRole('row')
      .filter({ has: this.page.getByRole('cell', { name: threadId, exact: true }) })
      .click();
    // Wait for thread panel to be visible
    await this.threadContainer.waitFor({ state: 'visible' });
  }

  async checkMessageInThread(message: string, isOutput: boolean = false): Promise<void> {
    if (isOutput) {
      await expect(this.outputContainer.filter({ hasText: message })).toBeVisible();
    } else {
      await expect(this.threadContainer.getByText(message)).toBeVisible();
    }
  }

  async searchForThread(threadId: string): Promise<void> {
    await this.searchInput.fill(threadId);
    await expect(this.threadRow).toHaveCount(1);
  }

  async deleteThreadFromTable(): Promise<void> {
    await this.threadCheckbox.click();
    await this.threadTableDeleteButton.click();
    await this.page.waitForTimeout(500);
    await this.threadDeletePopupDeleteButton.click();
  }

  async checkThreadIsDeleted(threadId: string): Promise<void> {
    await this.searchInput.fill(threadId);
    await expect(this.threadRow.filter({ hasText: threadId })).toHaveCount(0);
  }

  async deleteThreadFromThreadContentBar(): Promise<void> {
    await this.threadContainerDeleteButton.click();
    await this.threadDeletePopupDeleteButton.click();
  }

  async closeThreadContent(): Promise<void> {
    await this.page.keyboard.press('Escape');
  }
}
